### PR TITLE
Use the paths stored in the preview cache instead of copying the data over

### PR DESCRIPTION
**Pull request by kenxu95**
## Describe your changes and why you are making these changes
This should help the performance of the cache, especially when we start using slower storage like S3.

In order to accomplish this, I made the following two changes:
1) We check that the input paths exist before proceeding, throwing an appropriate error message. Ideally, use a different failure type like TRANSIENT which means the orchestrator can retry this issue. However, I don't expect this to be an issue since a cache hit should not be evicted that quickly.
2) We do not update the cache if the new entry is exactly the same as the existing one.

## Related issue number (if any)
ENG-1582

## Checklist before requesting a review
- [x] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [x] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [x] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: src/golang/lib/engine/engine.go [hunk 1]
```diff
@@ -15,7 +15,7 @@ import (
 const (
 	DefaultExecutionTimeout     = 15 * time.Minute
 	DefaultCleanupTimeout       = 2 * time.Minute
-	DefaultPollIntervalMillisec = 100
+	DefaultPollIntervalMillisec = 300
```
Comment (cw75) [inline src/golang/lib/engine/engine.go:18]: Is this change intentional?

File: src/golang/lib/workflow/operator/base.go [hunk 2]
```diff
@@ -113,18 +89,18 @@ func (bo *baseOperator) launch(ctx context.Context, spec job.Spec) error {
...
-				// We've successfully used the cached result!
-				return nil
+			// Apply the cached results for each output artifact. This just means setting the output paths
+			// to be the same as the cached ones.
+			for i, outputArtifact := range bo.outputs {
+				cacheEntry := cacheEntryByKey[outputArtifact.Signature()]
+				bo.outputExecPaths[i].ArtifactMetadataPath = cacheEntry.ArtifactMetadataPath
+				bo.outputExecPaths[i].ArtifactContentPath = cacheEntry.ArtifactContentPath
+
+				// The op metadata path is updated for both for this operator and its output artifacts.
```
Comment (cw75) [inline src/golang/lib/workflow/operator/base.go:99]: typo: "for both this operator and its output artifacts."

File: src/golang/lib/workflow/preview_cache/preview_cache.go [hunk 3]
```diff
@@ -82,19 +77,53 @@ func (c *inMemoryPreviewCacheManagerImpl) Put(ctx context.Context, artifactSigna
...
 			ArtifactContentPath:  execPathsList[i].ArtifactContentPath,
 			ArtifactMetadataPath: execPathsList[i].ArtifactMetadataPath,
 			OpMetadataPath:       execPathsList[i].OpMetadataPath,
 		})
+
+		// After adding the new value to the cache, delete the orphaned data if this overwrote an existing
+		// entry. This is best-effort.
+		if existingEntry != nil {
+			deleteDataForEntry(ctx, c.storageConfig, *existingEntry)
+		}
```
Comment (cw75) [inline src/golang/lib/workflow/preview_cache/preview_cache.go:126]: Will we ever reach here?

File: src/golang/lib/workflow/preview_cache/preview_cache.go [hunk 4]
```diff
@@ -82,19 +77,53 @@ func (c *inMemoryPreviewCacheManagerImpl) Put(ctx context.Context, artifactSigna
...
+		val, exists := c.cache.Peek(signatures)
+		if exists {
+			existingEntry = castCachedValueToEntry(val)
+			if existingEntry == nil {
+				return errors.New("Preview Artifact Cache is storing an unexpected data structure.")
+			}
+
+			// If the entry already exists and has the same data, short-circuit as there is no need
+			// to update anything.
+			if isEqual(execPathsList[i], existingEntry) {
```
Comment (cw75) [inline src/golang/lib/workflow/preview_cache/preview_cache.go:111]: If there is a cache hit, since we aren't copying these paths any more, isn't `execPathsList[i]` and `existingEntry` guaranteed to be equal? If so, I think we should do an assertion instead?

**Review by cw75 — COMMENTED** [review]
Looks good overall! Left a few comments inline. I'm fine with shipping this PR, but let's create a task that handles the retry part if the data read failed due to an unlikely event of cache eviction.

File: src/golang/lib/engine/engine.go (hunk 1 above)
Comment (kenxu95) [reply src/golang/lib/engine/engine.go:18]: yeah, I noticed during testing that we were polling super aggressively so I wanted to back that off a little bit.

File: src/golang/lib/workflow/preview_cache/preview_cache.go (hunk 4 above)
Comment (kenxu95) [reply src/golang/lib/workflow/preview_cache/preview_cache.go:111]: You are right that if there is a cache hit, these are guaranteed to be equal. However, this is the in the Put() part of the cache, which is called by artifact.Finish(), which has no idea whether the artifact was generated from a cache hit or not. 

We could move this logic into artifact, and only Put() if the artifact was generated from a cache hit, but I feel it belongs better at the cache layer.

File: src/golang/lib/workflow/preview_cache/preview_cache.go (hunk 3 above)
Comment (kenxu95) [reply src/golang/lib/workflow/preview_cache/preview_cache.go:126]: I don't think it will. Can drop a note here. But I do think it's totally reasonable to have this logic, since it means the cache doesn't rely on any assumptions about the cache user. Eg. if we change the cache entry to be mutable in the future, we won't have to remember to include this logic, since it's already here.

File: src/golang/lib/workflow/preview_cache/preview_cache.go (hunk 4 above)
Comment (cw75) [reply src/golang/lib/workflow/preview_cache/preview_cache.go:111]: I agree it should belong in the caching layer, but since this is already within the `if exists` block, which means we are in the cache hit case, I think we should change this to an assertion?

File: src/golang/lib/workflow/preview_cache/preview_cache.go (hunk 3 above)
Comment (cw75) [reply src/golang/lib/workflow/preview_cache/preview_cache.go:126]: Yeah let's drop a note here.

File: src/golang/lib/workflow/preview_cache/preview_cache.go (hunk 4 above)
Comment (kenxu95) [reply src/golang/lib/workflow/preview_cache/preview_cache.go:111]: gotcha, that makes sense. I'll throw an error since Go doesn't have assertions.

File: src/golang/lib/workflow/preview_cache/preview_cache.go (hunk 3 above)
Comment (kenxu95) [reply src/golang/lib/workflow/preview_cache/preview_cache.go:126]: actually on second thought, it looks better without this logic now that we error out if the entries are not equal. So its just less confusing if we remove it.

**Review by cw75 — APPROVED** [review]
